Write API: create proposals (#15790)
decidim/decidim: docs/modules/develop/pages/api/reference/components/proposals/create.adoc

= Create Proposal

== Input attributes
When you want to add or create data in your proposals component, you need to rely on the `ProposalAttributes` type, which has the following input fields:

```
• title       (required) - 15-150 characters
• body        (required) - minimum 15 characters
• address     (optional) - physical address (requires geocoding)
• latitude    (optional) - coordinate
• longitude   (optional) - coordinate
• taxonomies  (optional) - Taxonomies ids to assign to this proposal
```

== Sample request

The `CreateProposal` mutation allows authenticated users to create new proposals in a Decidim proposals component. This mutation uses the existing `CreateProposal` command from the controller.

[source,graphql]
----
mutation {
  component(id: "123") {
    ... on ProposalsMutation {
      createProposal(input: {
        attributes: {
          title: "Install Bike Lanes on Main Street",
          body: "We propose adding dedicated bike lanes along Main Street to improve cyclist safety and encourage sustainable transportation.",
          address: "Main Street, Barcelona, Spain",
          latitude: 41.3851,
          longitude: 2.1734,

>           "taxonomies": [123, 456]

Author: fixed by [1d8993b](https://github.com/decidim/decidim/pull/15790/commits/1d8993bccd2b4ebb9c4b6c9dc3f77016f27b9a5a)

>         }},

Author: fixed by [1d8993b](https://github.com/decidim/decidim/pull/15790/commits/1d8993bccd2b4ebb9c4b6c9dc3f77016f27b9a5a)

= Create Proposal

== Input attributes
When you want to add or create data in your proposals component, you need to rely on the `ProposalAttributes` type, which has the following input fields:

[source]
----
• title       (required) - 15-150 characters
• body        (required) - minimum 15 characters
• address     (optional) - physical address (requires geocoding)
• latitude    (optional) - coordinate
• longitude   (optional) - coordinate
• taxonomies  (optional) - Taxonomies ids to assign to this proposal
----

== Sample request

The `CreateProposal` mutation allows authenticated users to create new proposals in a Decidim proposals component. This mutation delegates to the existing `Decidim::Proposals::CreateProposal` command used by the controllers.

[source,graphql]
----
mutation {

>   component(id: "123") {
>     ... on ProposalsMutation {
>       createProposal(
>         input: {
>           attributes: {
>             title: "Install Bike Lanes on Main Street",
>             body: "We propose adding dedicated bike lanes along Main Street to improve cyclist safety and encourage sustainable transportation.",
>             address: "Main Street, Barcelona, Spain",
>             latitude: 41.3851,
>             longitude: 2.1734,
>             taxonomies: [123, 456]
>           },
>           locale: "en"
>         }
>       ) {
>         id
>         title {
>           translation(locale: "en")
>         }
>         address
>         coordinates {
>           latitude
>           longitude
>         }
>       }
>     }
>   }
> }
> ----

Author: ```suggestion
[source,graphql]
----
mutation createProposal($componentId: ID!, $input: CreateProposalInput!){
  component(id: $componentId) {
    ...on ProposalsMutation{
      createProposal(input: $input) {
        id
        title {
          translation(locale: "en")
        }
        body {          
          translation(locale: "en")
        }
        address
        taxonomies{
          name {
            translation(locale: "en")
          }
          children{
            name{
              translation(locale: "en")
            }
          }
        }
      }
    }
  } 
}
----

Example of submitted variables

[source,json]
----
{
  "componentId": 9,
  "input": {
    "locale": "en",
    "attributes": {
      "title": "Install Bike Lanes on Main Street",
      "body": "We propose adding dedicated bike lanes along Main Street to improve cyclist safety and encourage sustainable transportation.",
      "address": "Main Street, Barcelona, Spain",
      "latitude": 41.3851,
      "longitude": 2.1734
    }
  }
}
```

= Create Proposal

> == Input attributes

Reviewer: ```suggestion
== Input attributes

```

[source,graphql]
----
mutation createProposal($componentId: ID!, $input: CreateProposalInput!){
  component(id: $componentId) {
    ...on ProposalsMutation{
      createProposal(input: $input) {
        id
        title {
          translation(locale: "en")
        }
        body {          
          translation(locale: "en")
        }
        address
        taxonomies{
          name {
            translation(locale: "en")
          }
          children{
            name{
              translation(locale: "en")
            }
          }
        }
      }
    }
  } 
}
----

Example of submitted variables

[source,json]
----
{
  "componentId": 9,
  "input": {
    "locale": "en",
    "attributes": {
      "title": "Install Bike Lanes on Main Street",
      "body": "We propose adding dedicated bike lanes along Main Street to improve cyclist safety and encourage sustainable transportation.",
      "address": "Main Street, Barcelona, Spain",
      "latitude": 41.3851,
      "longitude": 2.1734
    }
  }

> }

Reviewer: ```suggestion
}
----
```
